### PR TITLE
BUG: pivot/unstack leading to too many items should raise exception

### LINKED ISSUE #20601: ERR: pivot_table when number of levels larger than int32 range
#### Code Sample, a copy-pastable example if possible

```
import pandas as pd
dat = pd.DataFrame({'ind1':list(range(1337600))*2,'ind2':list(range(3040))*2*440,'count':[1]*2*1337600})
dat.pivot_table(index='ind1',columns='ind2',values='count',aggfunc='count')
```

#### Problem description

Above code raises the following error:

```
  File "..\pandas\core\reshape\reshape.py", line 144, in _make_selectors
    mask = np.zeros(np.prod(self.full_shape), dtype=bool)

ValueError: negative dimensions are not allowed
```

np.prod(self.full_shape) appears to be returning a negative value because the number of unique index combinations is larger than the largest int32 value.

If line 144 were changed to the following, the issue could be fixed:
```
mask = np.zeros(np.prod(self.full_shape,dtype=np.int64), dtype=bool)
```

#### Expected Output
1337600 x 3040 dataframe.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.4.final.0
python-bits: 64
OS: Windows
OS-release: 7
machine: AMD64
processor: Intel64 Family 6 Model 37 Stepping 1, GenuineIntel
byteorder: little
LC_ALL: None
LANG: en
LOCALE: None.None

pandas: 0.22.0
pytest: 3.3.2
pip: 9.0.1
setuptools: 38.4.0
Cython: 0.27.3
numpy: 1.12.1
scipy: 1.0.0
pyarrow: None
xarray: None
IPython: 6.2.1
sphinx: 1.6.6
patsy: 0.5.0
dateutil: 2.6.1
pytz: 2017.3
blosc: None
bottleneck: 1.2.1
tables: 3.4.2
numexpr: 2.6.4
feather: None
matplotlib: 2.1.2
openpyxl: 2.4.10
xlrd: 1.1.0
xlwt: 1.3.0
xlsxwriter: 1.0.2
lxml: 4.1.1
bs4: 4.6.0
html5lib: 1.0.1
sqlalchemy: 1.2.1
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by sweb**
- [x] closes #20601
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by jreback** [comment]
is this like #20784 ? (you can take that over if you wish)

**Comment by sweb** [comment]
Yes, it is quite similar, however my changes are a bit more local. I do not attempt to compute how many rows/cols are allowed but just check whether the already computed value is negative due to an integer overflow.

However, it appears that this is OS dependent. On Unix `np.product` will not result in a negative value, but directly computes with `int64` or promotes it after reaching a limit.

In my current solution I explicitly set `dtype` of `np.product` to `int32` - this will mean a change in current behavior for Unix systems.

Do you think this is a valid approach or should we set it to `int64` as originally proposed?

**Comment by jreback** [comment]
@sweb the approach in the other issue is good.

**Comment by jreback** [comment]
you can simply cherry-pick that commit (ideally). or you can copy it whatever works.

**Comment by sweb** [comment]
@jreback Thanks for the advice. I patched the commits of the other PR into this branch. Unfortunately, the proposed solution does not solve the problem - at least not on my environment.

Since this is most likely a Windows-specific error (I checked the example from #20601 on MacOs and everything works fine - I did not check it on Linux yet) the proposed solution would actually limit the capabilities of unstack().

In addition, on my Win10 machine, I still get an error, since `num_rows * num_columns` also overflows and the condition `if num_rows * num_columns > (2 ** 31 - 1)` is never met.

I modified the PR in that a numpy warning, indicating the overflow is caught instead of checking the upper bound of int32. The tests are only applied to Windows environments.

Can you replicate this? Maybe my Windows environment setup is screwed up.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/reshape/test_pivot.py [hunk 1]
```diff
@@ -1276,6 +1277,18 @@ def test_pivot_string_func_vs_func(self, f, f_numpy):
                                aggfunc=f_numpy)
         tm.assert_frame_equal(result, expected)
 
+    @pytest.mark.slow
+    def test_pivot_number_of_levels_larger_than_int32(self):
+        # GH 20601
+        if sys.platform == 'win32':
```
[the remark this answers is not in the document]
Comment (sweb) [reply pandas/tests/reshape/test_pivot.py]: I don't know 100% percent, but numpy appears to treat int32 differently on Windows than on Unix:
https://stackoverflow.com/questions/36278590/numpy-array-dtype-is-coming-as-int32-by-default-in-a-windows-10-64-bit-machine/36279549

https://github.com/numpy/numpy/issues/8433

On my MacOs environment, the code from the issue works - takes some time but pandas (0.23.4) computes the expected result while my Windows environment encounters the described error. I therefore think, that this is a Windows only issue and we probably should not limit other OS versions by always checking against the int32 upper limit.

File: pandas/core/reshape/pivot.py [hunk 2]
```diff
@@ -81,9 +81,7 @@ def pivot_table(data, values=None, index=None, columns=None, aggfunc='mean',
                 pass
         values = list(values)
 
-    # group by the cartesian product of the grouper
-    # if we have a categorical
-    grouped = data.groupby(keys, observed=False)
```
[the remark this answers is not in the document]
Comment (sweb) [reply pandas/core/reshape/pivot.py]: Good question - this is from the original pull request and I did not know what it was for. I will change it back to the way it was.

File: pandas/core/reshape/reshape.py [hunk 3]
```diff
@@ -127,6 +127,16 @@ def __init__(self, values, index, level=-1, value_columns=None,
         self.removed_level = self.new_index_levels.pop(self.level)
         self.removed_level_full = index.levels[self.level]
 
+        num_rows = np.max([index_level.size for index_level
```
[the remark this answers is not in the document]
Comment (sweb) [reply pandas/core/reshape/reshape.py]: Good point - I will add the original comment from a later commit I did not cherry pick.

File: pandas/tests/reshape/test_pivot.py [hunk 4]
```diff
@@ -1276,6 +1277,18 @@ def test_pivot_string_func_vs_func(self, f, f_numpy):
...
+            df = DataFrame({'ind1': np.arange(2 ** 16),
+                            'ind2': np.arange(2 ** 16),
```
[the remark this answers is not in the document]
Comment (sweb) [reply pandas/tests/reshape/test_pivot.py]: Can you explain this further? If I decrease the value to `2 ** 15` I do not get an overflow.

**Comment by jreback** [comment]
can you merge master

**Comment by sweb** [comment]
The ValueError now raises regardless of OS.

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm. if you can add a whatsnew, put in reshaping bug fix section. saying that the error message is now improved, for pivoting with > int32.max uniques (obviously make a nice sentence about this).

**Comment by sweb** [comment]
Whatsnew added, thank you for your help!

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks!

**Pull request merged**